validation/manifest.go

```diff
@@ -10,21 +14,116 @@ import (
-	// 	return fmt.Errorf("manifest groups: %v", err)
-	// }
+	helper := validateManifestGroupsHelper{
+		hostnames: make(map[string]int),
+	}
+	names := make(map[string]int) // used as a set
+	for _, group := range groups {
+		if err := validateManifestGroup(group, &helper); err != nil {
+			return err
+		}
+		if _, exists := names[group.GetName()]; exists {
+			return fmt.Errorf("%w: duplicate group %q", ErrInvalidManifest, group.GetName())
+		}
+
+		names[group.GetName()] = 0 // Value stored is not used
+	}
+	if helper.globalServiceCount == 0 {
+		return fmt.Errorf("%w: zero global services", ErrInvalidManifest)
+	}
+	return nil
+}
+
+func validateManifestGroup(group manifest.Group, helper *validateManifestGroupsHelper) error {
+	if 0 == len(group.Services) {
+		return fmt.Errorf("%w: group %q contains no services", ErrInvalidManifest, group.GetName())
+	}
+
+	if err := dtypes.ValidateResourceList(group); err != nil {
+		return err
+	}
+	for _, s := range group.Services {
+		if err := validateManifestService(s, helper); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateManifestService(service manifest.Service, helper *validateManifestGroupsHelper) error {
+	if len(service.Name) == 0 {
+		return fmt.Errorf("%w: service name is empty", ErrInvalidManifest)
+	}
+
+	if len(service.Image) == 0 {
+		return fmt.Errorf("%w: service %q has empty image name", ErrInvalidManifest, service.Name)
+	}
+
+	for _, envVar := range service.Env {
+		idx := strings.Index(envVar, "=")
+		if idx == 0 {
+			return fmt.Errorf("%w: service %q defines an env. var. with an empty name", ErrInvalidManifest, service.Name)
+		}
+	}
+
+	for _, serviceExpose := range service.Expose {
+		if err := validateServiceExpose(service.Name, serviceExpose, helper); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateServiceExpose(serviceName string, serviceExpose manifest.ServiceExpose, helper *validateManifestGroupsHelper) error {
+	if serviceExpose.Port == 0 {
+		return ErrServiceExposePortZero
+	}
+
+	switch serviceExpose.Proto {
+	case manifest.TCP, manifest.UDP:
+		break
+	default:
+		return fmt.Errorf("%w: service %q protocol %q unknown", ErrInvalidManifest, serviceName, serviceExpose.Proto)
+	}
+
+	if serviceExpose.Global {
+		helper.globalServiceCount++
+	}
+
+	for _, host := range serviceExpose.Hosts {
+		if !isValidHostname(host) {
+			return fmt.Errorf("%w: service %q has invalid hostname %q", ErrInvalidManifest, serviceName, host)
+		}
+
+		_, exists := helper.hostnames[host]
+		if exists {
+			return errors.Errorf("hostname %q is duplicated, this is not allowed", host)
+		}
+		helper.hostnames[host] = 0 // Value stored does not matter
+	}
+
 	return nil
 }
 
-// ValidateManifestWithGroupSpecs does validation for manifest with group specifications
+var hostnameRegex = regexp.MustCompile(`^[[:alnum:],-,\.]+$`)
+
+func isValidHostname(hostname string) bool {
+	return hostnameRegex.MatchString(hostname)
```

Comment:
maybe also

* max length: 255
* min length: I don't know, 3?
* require host + . + tld?

It'd be cool if there was a hostname validation library that we could use.

Reply:
I did not find a library but I can put some additional limits on this.